### PR TITLE
Refactor Unit Collection

### PR DESCRIPTION
**Pull request by RoiEXLab**
This removes some duplicate code

### CODE REVIEW

File: src/test/java/games/strategy/engine/data/UnitCollectionTest.java [hunk 1]
```diff
@@ -142,7 +142,7 @@ public void unitCollection() {
   public void addUnit() {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/UnitCollectionTest.java]: Test method should be renamed `add()`.

File: src/test/java/games/strategy/engine/data/UnitCollectionTest.java [hunk 2]
```diff
@@ -154,7 +154,7 @@ public void addUnit() {
   public void addAllUnitsFromCollection() {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/UnitCollectionTest.java]: Test method should be renamed `addAllFromCollection()`.

File: src/test/java/games/strategy/engine/data/UnitCollectionTest.java [hunk 3]
```diff
@@ -166,8 +166,8 @@ public void addAllUnitsFromUnitCollection() {
     final int notifyChangedCountBefore = getDefaultPlayerNotifyChangedCounter();
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/UnitCollectionTest.java]: Test method should be renamed `addAllFromUnitCollection()`.

File: src/test/java/games/strategy/engine/data/UnitCollectionTest.java [hunk 4]
```diff
@@ -177,10 +177,10 @@ public void addAllUnitsFromUnitCollection() {
   @Test
   public void removeAllUnits() {
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/UnitCollectionTest.java]: Test method should be renamed `removeAll()`.

File: src/test/java/games/strategy/util/MatchTest.java [hunk 5]
```diff
@@ -109,11 +109,12 @@ public void testGetNMatches_ShouldThrowExceptionWhenMaxIsNegative() {
 
   @Test
   public void testAllMatch() {
```
Comment (ssoloff) [inline src/test/java/games/strategy/util/MatchTest.java]: I would recommend having separate test methods for `allMatch()` and `allMatchNotEmpty()`.  Just copy the original one and change the assertion for the empty collection scenario:

```java
@Test
public void testAllMatchNotEmpty() {
  assertFalse("empty collection", Match.allMatchNotEmpty(Arrays.asList(), IS_ZERO_MATCH));
  assertFalse("none match", Match.allMatchNotEmpty(Arrays.asList(-1, 1), IS_ZERO_MATCH));
  assertFalse("some match", Match.allMatchNotEmpty(Arrays.asList(-1, 0, 1), IS_ZERO_MATCH));
  assertTrue("all match (one element)", Match.allMatchNotEmpty(Arrays.asList(0), IS_ZERO_MATCH));
  assertTrue("all match (multiple elements)", Match.allMatchNotEmpty(Arrays.asList(0, 0, 0), IS_ZERO_MATCH));
}

@Test
public void testAllMatch() {
  assertTrue("empty collection", Match.allMatch(Arrays.asList(), IS_ZERO_MATCH));
  assertFalse("none match", Match.allMatch(Arrays.asList(-1, 1), IS_ZERO_MATCH));
  assertFalse("some match", Match.allMatch(Arrays.asList(-1, 0, 1), IS_ZERO_MATCH));
  assertTrue("all match (one element)", Match.allMatch(Arrays.asList(0), IS_ZERO_MATCH));
  assertTrue("all match (multiple elements)", Match.allMatch(Arrays.asList(0, 0, 0), IS_ZERO_MATCH));
}
```

I wouldn't worry about the appearance of duplicate test code here.  Tests are cheap and the documentation for specific methods is worth it in the long term.  Right now someone looking at this test would be hard-pressed to notice the one `allMatch()` in the sea of `allMatchNotEmpty()`s. :smile:

File: src/main/java/games/strategy/util/Match.java [hunk 6]
```diff
@@ -96,70 +89,45 @@ private Match(final Predicate<T> condition) {
 
   /**
    * returns true if all elements in the collection match.
+   * returns false if the collection is empty.
+   */
+  public static <T> boolean allMatchNotEmpty(final Collection<T> collection, final Match<T> match) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/Match.java]: Yeah... not sure what to think about this.  When I first saw it, I thought it smelled because we are now leaking an implementation detail that might possibly be a bug.  However, after reviewing some of the call sites, there are definitely some cases where having an empty collection should fail the match.

Since we are modifying 80+ call sites with this change, I was going to suggest we extract the `&& !collection.isEmpty()` expression into the larger expression at the call site where this method is called.  That would put it square in the face of the next dev who has to review those Boolean conditions so that they have to think of the implications of an empty collection.  In a sense, I want to discourage calls to this method because the empty collection test seems like a concern that `Match` should not be worried about.

So, the more I think about it, the more I'm ok with adding this method.  I would suggest it possibly be renamed to `allMatchAndNotEmpty()`, the `And` making it a little easier to read and putting the dev on notice that this is a compound condition (similar to how Guava named `Strings#isNullOrEmpty()`).  Actually, `notEmptyAndAllMatch()` probably makes more sense, but it then doesn't really follow the method naming conventions in this class.  But maybe that's what we want to make a dev take notice?

@ron-murhammer Since you looked at `allMatch()` returning `false` for an empty collection recently, do you have any thoughts?

File: src/main/java/games/strategy/util/Match.java [hunk 7]
```diff
@@ -96,70 +89,45 @@ private Match(final Predicate<T> condition) {
...
-    final Iterator<K> keys = map.keySet().iterator();
-    while (keys.hasNext()) {
-      final K key = keys.next();
-      final V value = map.get(key);
-      if (valueMatch.match(value)) {
-        rVal.add(key);
-      }
-    }
-    return rVal;
+    return map.keySet().stream().filter(key -> valueMatch.match(map.get(key))).collect(Collectors.toSet());
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/Match.java]: NBD, but you can avoid the lookup cost for each element by iterating over the entries instead of the keys:

```java
public static <K, V> Set<K> getKeysWhereValueMatch(final Map<K, V> map, final Match<V> valueMatch) {
  return map.entrySet().stream()
      .filter(entry -> valueMatch.match(entry.getValue()))
      .map(Map.Entry::getKey)
      .collect(Collectors.toSet());
}
```

File: src/main/java/games/strategy/engine/data/UnitCollection.java [hunk 8]
```diff
@@ -7,14 +7,15 @@
...
+import java.util.stream.Collectors;
 
 import games.strategy.util.IntegerMap;
 import games.strategy.util.Match;
 
 /**
  * A collection of units.
  */
-public class UnitCollection extends GameDataComponent implements Iterable<Unit> {
+public class UnitCollection extends GameDataComponent implements Collection<Unit> {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/UnitCollection.java]: I wasn't sure if this was going to cause a compatibility problem with old save games, but apparently interfaces are not serialized with the class descriptor (good to know).  I tried loading a 3635 save game, and it appeared to work fine.

File: src/main/java/games/strategy/engine/data/UnitCollection.java [hunk 9]
```diff
@@ -257,4 +219,37 @@ public String toString() {
...
+  @Override
+  public boolean retainAll(Collection<?> collection) {
+    return m_units.retainAll(collection);
+  }
+
+  @Override
+  public void clear() {
+    m_units.clear();
+    m_holder.notifyChanged();
+  }
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/UnitCollection.java]: All of these newly-added methods have no callers.  Nor do they have any unit tests. 😞   Is it worth adding them just to have `UnitCollection` extend `Collection<Unit>`?  I can't find any code in this PR that now relies on `UnitCollection` being a `Collection<Unit>`.

I think it's fine to change the method names in this class to align with those in `Collection`.  However, I think extending `Collection<Unit>` at this time may be a YAGNI.  Is there some near-future change you're considering where this will be required?

File: src/main/java/games/strategy/util/Match.java (hunk 7 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/util/Match.java]: We'll the lookup cost for hash implementations is pretty low, but good suggestion anyway.

File: src/main/java/games/strategy/engine/data/UnitCollection.java (hunk 9 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/UnitCollection.java]: Match.anyMatch etc. requires a collection

File: src/main/java/games/strategy/util/Match.java (hunk 6 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/util/Match.java]: I agree, this method shouldn't be used.
But there are just too many calls to this method to check if the individual actually need the empty check. I'd suggest moving this change to another PR and making this method deprecated in the meantime. 
(I'm not going to rename it either, because I'm planning to remove it then.)

Comment (RoiEXLab) [inline src/main/java/games/strategy/util/Match.java]: Implementing collections also enables us to get rid of the `addAll(UnitCollection)` overload

File: src/main/java/games/strategy/engine/data/UnitCollection.java (hunk 9 above)
Comment (ssoloff) [inline src/main/java/games/strategy/engine/data/UnitCollection.java]: Sure, but I can't find any instances where we're now passing a `UnitCollection` to these methods.  Is the plan to remove the `Match`-related methods in `UnitCollection` and replace those with calls to the same-named methods in `Match`?  For example, instead of calling `unitCollection.allMatch(match)` you'd call `Match.allMatch(unitCollection, match)`?

File: src/test/java/games/strategy/engine/data/UnitCollectionTest.java [hunk 10]
```diff
@@ -139,52 +125,44 @@ public void unitCollection() {
...
+    unitCollection.addAll(unitsOfOtherPlayerOfUnitTypeOne);
 
-    final int notifyChangedCountBefore = getDefaultPlayerNotifyChangedCounter();
-    unitCollection.removeAllUnits(unitsOfOtherPlayerOfUnitTypeOne);
+    unitCollection.removeAll(unitsOfOtherPlayerOfUnitTypeOne);
 
     assertThat(unitCollection.getUnitCount(), is(equalTo(0)));
-    final int expNotifyChangedCountAfter = notifyChangedCountBefore + 1;
-    assertThat(getDefaultPlayerNotifyChangedCounter(), is(equalTo(expNotifyChangedCountAfter)));
+    verify(defaultPlayerId, times(2)).notifyChanged();
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/data/UnitCollectionTest.java]: In this case, the first invocation is a side effect of the "arrange/given" part of this test.  I don't like using `Mockito#reset()` in general, but this seems like it might be more appropriate than possibly misleading a future test reader that two calls to the handler are expected.  For example:

```java
@Test
public void removeAll() {
  final Collection<Unit> unitsOfOtherPlayerOfUnitTypeOne = getOtherPlayerUnitsOfUnitTypeOne();
  unitCollection.addAll(unitsOfOtherPlayerOfUnitTypeOne);
  reset(defaultPlayerId); // <-- ignore handler call from addAll()

  unitCollection.removeAll(unitsOfOtherPlayerOfUnitTypeOne);

  assertThat(unitCollection.getUnitCount(), is(equalTo(0)));
  verify(defaultPlayerId).notifyChanged(); // <-- verify only one call to handler from removeAll()
}
```

This [SE answer](https://softwareengineering.stackexchange.com/a/188604/189282) suggests that this type of `reset()` usage is ok, but that it would be preferable to have its use hidden in a separate method responsible for setting up the test, e.g. `givenUnitCollectionWithUnitsOfOtherPlayerOfUnitTypeOne()`.  That may not be necessary in this one case given that the arrange/act/assert parts of the test are already delineated by whitespace.
```

File: src/main/java/games/strategy/engine/data/UnitCollection.java (hunk 9 above)
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/data/UnitCollection.java]: I'm not a huge fan of the various wrapper types for collections of a specific type like IntegerMap or UnitCollection.
While IntegerMap is kinda useful because it adds operations like `add(key)` etc. UnitCollection is essentially a wrapper for an ArrayList of unit types wich notifies pseudo-observers whenever a value is being added or removed

**Comment by RoiEXLab** [comment]
@ssoloff Updated!

**Comment by ssoloff** [comment]
Ok, all of my comments have been addressed.  The 80+ new deprecation warnings are a bit annoying, but hopefully the new `Match#allMatchNotEmpty()` method will be removed soon.

Kicking over to @ron-murhammer to give him a chance to comment on the [original question](https://github.com/triplea-game/triplea/pull/2068#discussion_r125805003) posed to him before merging.

File: src/main/java/games/strategy/util/Match.java (hunk 6 above)
Comment (ron-murhammer) [inline src/main/java/games/strategy/util/Match.java]: Yeah this is a tough one. I'm alright with this solution for now.
